### ISSUE TITLE
Incompatible with Webpack 2.1.0-beta.23 configuration validation

### ISSUE THREAD

**Issue by kylecesmat** [postcss/postcss-loader#99]
The latest webpack beta now validates the configuration object, which breaks this usage as the `postcss` key in the configuration is not part of the validated schema. See release notes here: https://github.com/webpack/webpack/releases/tag/v2.1.0-beta.23

```
module.exports = {
    module: {
        loaders: []
    },
    postcss: function () {}
}
```

The entire output is:
```
WebpackOptionsValidationError: Invalid configuration object. Webpack has been initialised using a configuration object that does not match the API schema.
 - configuration has an unknown property 'postcss'. These properties are valid:
   object { amd?, bail?, cache?, context?, devServer?, devtool?, entry, externals?, loader?, module?, name?, dependencies?, node?, output?, plugins?, profile?, recordsInputPath?, recordsOutputPath?, recordsPath?, resolve?, resolveLoader?, stats?, target?, watch?, watchOptions? }
```

I understand this is for a webpack 2 beta, and probably should wait until the features stabilize. Who knows if this configuration validation will land in the final version...

For now I've locked webpack@2.1.0-beta.21

**Comment by ai** [comment]
I will start this feature after we will stabilize config https://github.com/postcss/postcss-loader/pull/97

**Comment by donaldpipowitch** [comment]
Would this work as an ad-hoc fix? https://github.com/webpack/webpack/pull/2974#issuecomment-245857168

**Comment by phaistonian** [comment]
@donaldpipowitch I am also curious. Would using this approach enable us use postCSS on the latest beta?

**Comment by wellyshen** [comment]
Same issue +1

**Comment by SpaceK33z** [comment]
See https://github.com/webpack/webpack/issues/3018#issuecomment-248187254 for a temporary workaround, which should work without a fix in `postcss-loader`. Note that this is meant to be temporary, because it is now possible to configure the options in the `query` object (but it looks like you already know this).

**Comment by kevinSuttle** [comment]
So, are the Webpack options for postcss or the postcss-plugins?

**Comment by MoOx** [comment]
We must implement this asap. I can handle it, this change is required for me to cut a release for phenomic which require webpack 2.
I will try to push a PR quickly. It should be easy to support webpack 1 & 2 at the same time since query can know accept normal JS (so function).

**Comment by MoOx** [comment]
Well according to #97 I am going to way. @ai I think with webpack 2 something could be done to support both version of webpack. I wrote several webpack loaders so poke me if you need a hand ;)

**Comment by MoOx** [comment]
Problem is, the workaround mentionned earlier is not working if you use css modules. @ai poke me when you are ready to more forward, this is a small "big deal" we have to handle here :D

**Comment by ai** [comment]
@MoOx do we have any webpack 2 loader API changelog?

What changes do you suggest?

**Comment by MoOx** [comment]
Changelog is: you can't rely on "random" properties in webpack global config.
So what you use to call via `this.options.postcss` have to be replaced.
The good news is: your `params` (the "query"  associated with the loader) now accept real JS, not just json (like it used to). So we "simply" need to normalize options passed to the loader by only using the "query". This will basically deprecate the "pack" option since it won't be necessary anymore.

tl;dr: no more this.options.postcss and everything should be in `params`.

For backward compat we can refactor the code and simply read options like this

```
const options = {
   ...{ some default ? },
   ...this.options.postcss,  // backward compat with v1
   ...loaderUtils.parseQuery(this.query)
}
```

It's what I am doing for eslint-loader and it's working like a charm https://github.com/MoOx/eslint-loader/blob/f965eecfd5dac6489abeaebeb8fc7e5fa010bade/index.js#L128-L136

**Comment by Oopscurity** [comment]
Validation succeeds for me:

```javascript
plugins: [
  new webpack.LoaderOptionsPlugin({
    options: {
      context: __dirname,
      postcss: [
        autoprefixer
      ]
    }
  }
]
```

Where `autoprefixer` is

```javascript
const autoprefixer = require('autoprefixer');
```

See more at https://github.com/webpack/webpack/issues/3018#issuecomment-248633498

**Comment by kevinSuttle** [comment]
@Oopscurity Can you not use `import` ?

**Comment by MoOx** [comment]
You can use it, by passing the function as you used too.

**Comment by kevinSuttle** [comment]
So it beta 25 the stable one now?

**Comment by dkrutsko** [comment]
@Oopscurity: I was having an issue with extract-loader as well because it couldn't find the `output.publicPath` definition. So in addition to `context,` you also need to provide any additional options for the loaders you are using. I'm guessing `webpack.LoaderOptionsPlugin` creates a localized webpack configuration environment.

**Comment by kevinSuttle** [comment]
Following up on: https://github.com/postcss/postcss-loader/issues/99#issuecomment-248388686 

Before, there was [this method](https://github.com/postcss/postcss-custom-properties#usage):
```js
// dependencies
var fs = require("fs")
var postcss = require("postcss")
var customProperties = require("postcss-custom-properties")

// css to be processed
var css = fs.readFileSync("input.css", "utf8")

// process css using postcss-custom-properties
var output = postcss()
  .use(customProperties())
  .process(css)
  .css
```

but now the API has changed. I'm struggling to figure out why there are so many. 

**webpack config**
``` js
new webpack.LoaderOptionsPlugin({
        options: {
          postcss: function (webpack) {
            return [
            importCSS({
              addDependencyTo: webpack
            }),
              customProperties({
                   preserve: true,
                   appendVariables: true,
              }),
              cssApply,
              autoprefixer,
            ]
          }
        }
      }),
```

**Directory structure**
- webpack.config.js
- /src/Buttons/Button.jsx 
    - (contains a `require(./Buttons.css)`, 
    - which relies on the custom properties defined in ./src/Styles/Index.css
    - which are imported from ./node_modules/ (

Yet I still get warnings the custom properties as used without fallbacks, _unless_ I `@import ./Styles/Index.css` into every single component's css file (e.g. Buttons.css). 

That solves one problem, but then introduces another:
Now every single component has the imports inlined, bloating up the CSS.

I've noticed that there are multiple style tags too, but I'm not sure why that's happening.
I suppose it's because I'm not using Extract-Text. 

I have a repo I can send privately if it helps.

**Comment by MoOx** [comment]
@kevinSuttle This have NOTHING to do with how postcss-loader works, or webpack 2 update. This is why you have a way to define variables directly from your customProperties config. See readme of postcss-custom-properties.

**Comment by ai** [comment]
Fix by https://github.com/postcss/postcss-loader/pull/104

Now we have `plugins` option to query.

**Comment by ai** [comment]
I finished all changes and ready for release. Could somebody test this plugin on real application? (Just replace version in `package.json` to `postcss/postcss-loader`)

**Comment by kevinSuttle** [comment]
I don't see a newer version than 0.13.0 (the one I have now)?

**Comment by SpaceK33z** [comment]
@kevinSuttle, as @ai said, you need replace `postcss-loader` temporarily in `package.json` with this:

```
"postcss-loader": "postcss/postcss-loader"
```

**Comment by kevinSuttle** [comment]
Works!

**Comment by farwayer** [comment]
@kevinSuttle I opened issue https://github.com/michael-ciniawsky/postcss-load-config/issues/24

**Comment by michael-ciniawsky** [comment]
fix in [postcss-load-config v1.0.0-beta](https://github.com/michael-ciniawsky/postcss-load-config)

**Comment by ai** [comment]
Guys! Thanks for you tests.

Now we fixed most of them. Could you test `postcss-loader` from GitHub again?

**Comment by farwayer** [comment]
I'm afraid config from readme is not working:
```js
{
    loader: 'postcss-loader',
    options: {
        plugins: function () {
            return [
                require('precss'),
                require('autoprefixer')
            ];
        }
    }
}
```
Please read https://github.com/webpack/webpack/issues/3018#issuecomment-253336540.

**Comment by ai** [comment]
@farwayer hm, very strange it works for one users, and doesn’t work for rest :-/

**Comment by farwayer** [comment]
@ai syntax with loader options or with `LoaderOptionsPlugin`?

**Comment by farwayer** [comment]
Something wrong with `style-loader` https://github.com/webpack/webpack/issues/3018#issuecomment-253361436

**Comment by ai** [comment]
OK. @michael-ciniawsky publish `postcss-load-confg` release candidate. Let’s check `postcss-loader` last time.

**Comment by farwayer** [comment]
Loader options configurations will not work with `style-loader` but it's webpack core issue so nothing to do. Maybe it is better to keep note about this in readme before https://github.com/webpack/webpack/issues/3136 will be fixed.

**Comment by ai** [comment]
1.0 was released. Thanks for help.

**Comment by fmal** [comment]
@ai i've updated to 1.0 and followed instructions in readme for webpack 2.x config (i'm using latest webpack beta.25). Due to issue mentioned by @farwayer i couldn't make it work. Providing options via ` webpack.LoaderOptionsPlugin` solved the problem.

```js
new webpack.LoaderOptionsPlugin({
  options: {
    postcss () {
      return [ ... ];
    }
  }
});
```

Please consider updating the readme, so that people using beta builds of webpack don't waste time running into this.

**Comment by ai** [comment]
@fmal postcss-loader 1.0 doesn't need LoaderOptionsPlugin.

Could you show your config, I will try to find issue there.

**Comment by farwayer** [comment]
@ai add `style-loader` to loader chain and configuration via options will not work. Please check [this issue](https://github.com/webpack/webpack/issues/3136). Of course it is not postcss-loader bug but using postcss with style is very often used configuration in dev.
So adding warning to readme is very desirable.

**Comment by Birowsky** [comment]
> @fmal postcss-loader 1.0 doesn't need LoaderOptionsPlugin.

@ai I couldn't a guide for how to specify plugins without creating a new file. Help?

**Comment by ai** [comment]
@Birowsky define them in `package.json`: https://github.com/michael-ciniawsky/postcss-load-config#packagejson

**Comment by Birowsky** [comment]
@ai I'm not a fan of that approach either. Can't the `ExtractTextPlugin.extract({..})` function receive the plugins in the object?

**Comment by Birowsky** [comment]
@ai I prefer keeping all the config inside the webpack config file. So I'll stick with the `LoaderOptionsPlugin`. Thanx anyways.

**Comment by michael-ciniawsky** [comment]
@Birowsky 

webpack: >v2.0.0-beta.25

**webpack.config.js**
```js
module: {
rules : [{
   test: /\.(sass|less|styl|sss|css)$/,
   use: [
        ...// style-loader, css-loader
        {
          loader: 'postcss-loader',
          options: { options: { /* PostCSS Options */ }, plugins: () => [ /* PostCSS Plugins */ ]  }
        }
       ...// less-loader, sass-loader, stylus-loader
    ]
  }]
}
```

Basically like in LoaderOptionsPlugin, the plugin was more of an 'polyfill' for webpack 2 < beta.25, webpack 2 development is still a bit in flux :)

**Comment by felixsanz** [comment]
@michael-ciniawsky How to use that inline syntax? If i don't use a postcss.config.js file i'm getting

```
PostCSS Config could not be loaded. Please check your PostCSS Config.
PostCSS Config could not be loaded. Please check your PostCSS Config.
PostCSS Config could not be loaded. Please check your PostCSS Config.
```

I also would like to use the `options` and `plugins: ()` syntax.

**Comment by danieleleoni** [comment]
@michael-ciniawsky 
I have the same problem. I think it is due to this breaking change:
https://github.com/webpack/webpack/releases/tag/v2.1.0-beta.26

For the moment I have not a solution

**Comment by farskid** [comment]
@Oopscurity Your solution worked for me on webpack@2.2.1
